### ISSUE TITLE
Using timeout with some onTimeout closures on a stream subscription's cancel future throws exception

### ISSUE THREAD

**Issue by hampuslavin** [dart-lang/sdk#56846]
Hi! 👋 
I stumbled upon a strange case when working with subscriptions returned from `stream.listen`:

```dart
import 'dart:async';

void main() {
  var stream = StreamController<int>();

  var subscription = stream.stream.listen((event) {});

  subscription.cancel().timeout(Duration(seconds: 1), onTimeout: () async {
    // This closure's type is causing a runtime exception
  });
}
```

Running the above code will throw the following runtime exception:
```
Unhandled exception:
type '() => Future<void>' is not a subtype of type '(() => Future<Null>?)?' of 'onTimeout'
#0      Future.timeout (dart:async/future_impl.dart:939:54)
```

Am I missing something obvious or is this a bug? 

Dart version and tooling diagnostic info (`dart info`):
- Dart 3.5.3 (stable) (Wed Sep 11 16:22:47 2024 +0000) on "macos_arm64"
- on macos / Version 14.6.1 (Build 23G93)

Thanks!

**Comment by lrhn** [comment]
> Am I missing something obvious ...

I think it's pretty obvious that this heavily deprecated [declaration](https://github.com/dart-lang/sdk/blob/9a6503b5006b34eb5645f952f4b3e08dd8641f97/sdk/lib/internal/internal.dart#L160) hidden in an internal library that nobody will admit exists, has type `Future<Null>` and that that will break your program. Right?
Clear as rain!

Anyway, maybe it should have type `Future<void>` instead, so we can go on pretending it doesn't exist.
Let me try :wink:

**Comment by lrhn** [comment]
Fix [landed](https://dart-review.googlesource.com/c/sdk/+/388400). Hope it sticks. 
Don't know which version it will be shipped in, whether it can make it into SDK 3.6.0 or it has to wait for 3.7.0.

**Comment by hampuslavin** [comment]
Thanks for the quick turnaround! 🚀 🙌
